src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,197 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.using System;
+
+using System;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Sockets.Internal.Formatters
+{
+    public class ServerSentEventsMessageParser
+    {
+        const byte ByteCR = (byte)'\r';
+        const byte ByteLF = (byte)'\n';
+        const byte ByteT = (byte)'T';
+        const byte ByteB = (byte)'B';
+        const byte ByteC = (byte)'C';
+        const byte ByteE = (byte)'E';
+
+        private static byte[] _dataPrefix = Encoding.UTF8.GetBytes("data: ");
+        private static byte[] _sseLineEnding = Encoding.UTF8.GetBytes("\r\n");
+        private static byte[] _newLine = Encoding.UTF8.GetBytes(Environment.NewLine);
+
+
+        private InternalParseState _internalParserState = InternalParseState.ReadMessageType;
+        private List<byte[]> _data = new List<byte[]>();
+
+        public ParseResult ParseMessage(ReadableBuffer buffer, out ReadCursor consumed, out ReadCursor examined, out Message message)
+        {
+            consumed = buffer.Start;
+            examined = buffer.End;
+            message = new Message();
+            var reader = new ReadableBufferReader(buffer);
+
+            var start = consumed;
+            var end = examined;
+
+            while (!reader.End)
+            {
+                var messageType = MessageType.Text;
```

Comment:
You need to store this in a field, because if we see `data: T\r\n` in one buffer and `data: Foo\r\n` in the next buffer, we need to store the message type between invocations of the Parse method.